### PR TITLE
Allow total memory to be overridden

### PR DESCRIPTION
**Pull request by droberts195**
Since #65905 Elasticsearch has determined the Java heap settings
from node roles and total system memory.

This change allows the total system memory used in that calculation
to be overridden with a user-specified value. This is intended to
be used when Elasticsearch is running on a machine where some other
software that consumes a non-negligible amount of memory is running.
For example, a user could tell Elasticsearch to assume it was
running on a machine with 3GB of RAM when actually it was running
on a machine with 4GB of RAM.

The system property is `es.total_memory_bytes`, so, for example,
could be specified using `-Des.total_memory_bytes=3221225472`.
(It is specified in bytes rather than using a unit, because it
needs to be parsed by startup code that does not have access to
the utility classes that interpret byte size units.)

### CODE REVIEW

File: distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/MachineDependentHeap.java [hunk 1]
```diff
@@ -66,15 +66,32 @@ public MachineDependentHeap(SystemMemoryInfo systemMemoryInfo) {
...
+            .findFirst()
+            .orElse(null);
+        if (totalMemoryBytesOption == null) {
+            return null;
+        }
+        try {
+            return Long.parseLong(totalMemoryBytesOption.split("=", 2)[1]);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Unable to parse number of bytes from [" + totalMemoryBytesOption + "]");
+        }
```
Comment (pugnascotia) [inline distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/MachineDependentHeap.java:85]: I think you don't need an intermediate variable if you used `Optional.map`?

```suggestion
        return userDefinedJvmOptions.stream()
            .filter(option -> option.startsWith("-Des.total_memory_bytes="))
            .findFirst()
            .map(totalMemoryBytesOption -> {
	            try {
	                return Long.parseLong(totalMemoryBytesOption.split("=", 2)[1]);
	            } catch (NumberFormatException e) {
	                throw new IllegalArgumentException("Unable to parse number of bytes from [" + totalMemoryBytesOption + "]");
	            }
            })
            .orElse(null);
```

File: docs/reference/cluster/stats.asciidoc [hunk 2]
```diff
@@ -916,6 +916,18 @@ Total amount of physical memory across all selected nodes.
 (integer)
 Total amount, in bytes, of physical memory across all selected nodes.
 
+`total_override`::
+(<<byte-units,byte value>>)
+If the amount of physical memory has been overridden using the `es.total_memory_bytes`
+system property on all selected nodes then this reports the sum of the overridden
```
Comment (pugnascotia) [inline docs/reference/cluster/stats.asciidoc:922]: What if the total memory was only overridden on some of the selected nodes?

**Review by mark-vieira — APPROVED** [review]
I left a few comments. I think we should get somebody from core/infra to give this a once over as well. Otherwise :+1: .

**Review by pugnascotia — APPROVED** [review]

**Review by rjernst — COMMENTED** [review]
I left a couple comments. In general I think we should have this value replace the existing total memory, not be an additional stat next to it. In practice I think that means we need to take the difference between the override and real memory to adjust free memory.

**Review by breskeby — APPROVED** [review]
nothing to add besides already raised points. more reviewed this for learning about launcher logic

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
@elasticmachine update branch

**Comment by droberts195** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Pull request merged**